[6.2] Serve site modules in ModuleHelper when in API context

itsnewtjam opened the pull request:
### Summary of Changes

This change modifies ModuleHelper to process site modules when running in ApiApplication context. Currently the `/content/articles/{id}` Web Services endpoint prepares the article content (e.g. `{loadmodule}` tags), but since ModuleHelper filters by the Client ID, no modules are rendered under ApiApplication.

This is a very basic way to address this, open to going about it a different way if there are implications I didn't think of!

### Testing Instructions

Retrieve an article with modules rendered (assuming article ID 1 uses a load module tag of some sort)
GET /api/index.php/v1/content/articles/1

- Expected: Returns article data with modules rendered in the article text

### Actual result BEFORE applying this Pull Request

Despite the endpoint preparing article content, ModuleHelper filters modules by Client ID, so no modules are rendered under ApiApplication.

### Expected result AFTER applying this Pull Request

ModuleHelper serves site modules when under ApiApplication, resulting in modules being rendered in article content.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

muhme commented:
@itsnewtjam Thank you for your contribution :+1: Since 5.4.0 is already in feature freeze mode, this PR needs to be moved to 6.1-dev. If you could make the rebase yourself, that would be great. Otherwise, we can support it.

itsnewtjam committed:
serve site modules under api context

HLeithner commented:
This pull request has been automatically rebased to 6.2-dev.

laoneo committed:
Merge branch '6.2-dev' into api-site-modulehelper

QuyTon commented on libraries/src/Helper/ModuleHelper.php:
```diff
@@ -406,6 +407,10 @@ public static function getModuleList()
         $clientId = (int) $app->getClientId();
+        if ($clientId === 3)
+        {
+            $clientId = 0;
```
```suggestion
        if ($clientId === 3) {
            $clientId = 0;
```

MacJoom committed:
Update libraries/src/Helper/ModuleHelper.php

cs

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

ssnobben commented:
DId this get tested and merged?

alikon commented:
no, it still need test

alikon commented:
i've got  
<img width="617" height="261" alt="image" src="https://github.com/user-attachments/assets/e95a6644-8a2a-4cd4-9a5c-e694f3553f5c" />

so no html from custom module

alikon commented:
I have tested this item :red_circle: unsuccessfully on 605b25d06f2a4d4b2939c4f420badf1e768914ba<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46107">issues.joomla.org/tracker/joomla-cms/46107</a>.</sub>

QuyTon commented on libraries/src/Document/JsonapiDocument.php:
```diff
@@ -127,6 +127,21 @@ public function __toString()
+     *
+     * @return  RendererInterface
+     *
+     * @since   6.2.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

itsnewtjam commented:
So the solution originally used in this PR is now being cut off by #45357, but then in `api/components/com_content/src/View/Articles/JsonapiView.php` it's specifically dispatching onContentPrepare when preparing article items for the API.

https://github.com/joomla/joomla-cms/blob/b12cc3870625b81f52a6e07237f79d4c8db0ef0c/api/components/com_content/src/View/Articles/JsonapiView.php#L190-L200

Given other changes since I first created this PR, I've modified the approach. I've restored the ModuleHelper, removed that check in the LoadModule plugin, and created an override for `loadRenderer` in JsonapiDocument to set the renderer document type to HTML. Those two changes seem to now be sufficient to serve the module content.
